cgrep.py

```diff
@@ -0,0 +1,443 @@
+    # list[0] returned by get_nets
+    first = get_nets([first_obj], db)[0][1]
+    second = get_nets([second_obj], db)[0][1]
+
+  elif obj_type == "service":
+    first = get_ports([first_obj], db)[0][1]
+    second = get_ports([second_obj], db)[0][1]
+  else:
+    raise ValueError("check_encapsulated() currently only supports "
+                     "'network' and 'service' for the obj_type parameter")
+  # iterates over each object in the first group, and then each obj in the
+  # second group, making sure each one in the first is contained
+  # somewhere in the second.
+  for obj in first:
+    for sec_obj in second:
+      if obj in sec_obj:
+        break
+    # if we got through every object in the second group, and didn't have
+    # a match, then the first group is not entirely contained.
+    else:
+      return False
+  # if we got here, then the group was fully contained.
+  return True
+
+
+def print_diff(ip, common, diff1, diff2):
+  """ Print out the common, added, and removed network objects between 2 IPs
+
+  :ip str: the IP being compared against
+  :common list: the network objects shared between the two IPs
+               ('ip' and the other passed into options.cmp)
+  :diff1 list: the network objects present in 'ip' but not in the other IP
+               passed into options.cmp
+  :diff2 list: the network objects not present in 'ip' but are present in the
+               other IP passed into options.cmp
+
+  :returns None:
+  """
+  print("IP: %s" % ip)
+  if common:
+    common = ['  {0}'.format(elem) for elem in common]
+    print('\n'.join(common))
+  if diff1:
+    diff = ['+ {0}'.format(elem) for elem in diff1]
+    print('\n'.join(diff))
+  if diff2:
+    diff = ['- {0}'.format(elem) for elem in diff2]
+    print('\n'.join(diff))
+
+
+def group_diff(options, db):
+  """ Diffs two different group objects.
+
+  :options ArgumentParser.parse_args(): the options sent to the script
+  :db naming.Naming(): network and service definitions
+  :returns tuple: three lists, the common lines, the differences from 1 to 2,
+          and the differences from 2 to 1.
+  """
+  nested_rvals = []
+  for ip in options.gmp:
+    nested_rvals.append(get_ip_parents(ip, db))
+  # get just the list of groups, stripping out the networks.
+  group1 = [x[0] for x in nested_rvals[0]]
+  group2 = [x[0] for x in nested_rvals[1]]
+  common = list(set(group1) & set(group2))
+  diff1 = list(set(group1) - set(group2))
+  diff2 = list(set(group2) - set(group1))
+  return common, diff1, diff2
+
+
+def get_ip_parents(ip, db):
+  """ Gets a list of all network objects that include an IP.
+
+  :ip str: the IP we're looking for the parents of.
+  :db naming.Naming(): network and service definitions
+  :returns list: a list of all groups that include the IP, in the format:
+           [("Group", ["networks", "matched"]), (etc)]
+  """
+  results = []
+  rval = db.GetIpParents(ip)
+  for v in rval:
+    nested = db._GetParents(v, db.networks)
```

Comment:
Internally, we had developed a slightly different version of _GetParents (with a slightly different name) and called it directly instead of also changing GetNetParents to call it. When ripping that out, I reverted it back to just _GetParents out of habit.
Will update it.